cmd/systray/systray.go

```diff
@@ -64,39 +66,97 @@ func onReady() {
 
 	chState = make(chan ipn.State, 1)
 
+	menu := new(Menu)
+	menu.rebuild(fetchState(ctx))
+
+	go watchIPNBus(ctx)
+}
+
+type state struct {
+	status      *ipnstate.Status
+	curProfile  ipn.LoginProfile
+	allProfiles []ipn.LoginProfile
+}
+
+func fetchState(ctx context.Context) state {
 	status, err := localClient.Status(ctx)
 	if err != nil {
 		log.Print(err)
 	}
-
-	menu := new(Menu)
-	menu.rebuild(status)
-
-	go watchIPNBus(ctx)
+	curProfile, allProfiles, err := localClient.ProfileStatus(ctx)
+	if err != nil {
+		log.Print(err)
+	}
+	return state{
+		status:      status,
+		curProfile:  curProfile,
+		allProfiles: allProfiles,
+	}
 }
 
 // rebuild the systray menu based on the current Tailscale state.
 //
 // We currently rebuild the entire menu because it is not easy to update the existing menu.
 // You cannot iterate over the items in a menu, nor can you remove some items like separators.
 // So for now we rebuild the whole thing, and can optimize this later if needed.
-func (menu *Menu) rebuild(status *ipnstate.Status) {
+func (menu *Menu) rebuild(state state) {
 	menu.mu.Lock()
 	defer menu.mu.Unlock()
 
 	if menu.eventCancel != nil {
 		menu.eventCancel()
 	}
-	menu.status = status
+	ctx := context.Background()
+	ctx, menu.eventCancel = context.WithCancel(ctx)
+
+	menu.status = state.status
 	systray.ResetMenu()
 
 	menu.connect = systray.AddMenuItem("Connect", "")
 	menu.disconnect = systray.AddMenuItem("Disconnect", "")
 	menu.disconnect.Hide()
 	systray.AddSeparator()
 
-	if status != nil && status.Self != nil {
-		title := fmt.Sprintf("This Device: %s (%s)", status.Self.HostName, status.Self.TailscaleIPs[0])
+	account := "Account"
+	if state.curProfile.Name != "" {
+		account += fmt.Sprintf(" (%s)", state.curProfile.Name)
+	}
+	accounts := systray.AddMenuItem(account, "")
+	// The dbus message about this menu item must propagate to the receiving
+	// end before we attach any submenu items. Otherwise the receiver may not
+	// yet record the parent menu item and error out.
+	//
+	// On waybar with libdbusmenu-gtk, this manifests as the following warning:
+	//    (waybar:153009): LIBDBUSMENU-GTK-WARNING **: 18:07:11.551: Children but no menu, someone's been naughty with their 'children-display' property: 'submenu'
+	time.Sleep(100 * time.Millisecond)
+	// Aggregate all clicks into a shared channel.
+	menu.accountsCh = make(chan ipn.ProfileID)
+	for _, profile := range state.allProfiles {
+		title := fmt.Sprintf("%s (%s)", profile.Name, profile.NetworkProfile.DomainName)
+		// Note: we could use AddSubMenuItemCheckbox instead of this formatting
+		// hack, but checkboxes don't work across all desktops unfortunately.
+		if profile.ID == state.curProfile.ID {
+			title = "* " + title
```

Comment:
for future improvement, here is what I was doing to display tailnet name only when the profile name was ambiguous, as well as to display profile photos inline: https://github.com/tailscale/tailscale/blob/will-systray/cmd/systray/systray.go#L197

somewhere I had an alternate implementation that did caching of the http responses, but I'll have to find it.

Reply:
heh, actually it looks like that does have the caching code in https://github.com/tailscale/tailscale/blob/will-systray/cmd/systray/systray.go#L263-L280, but is also doing the non-caching version elsewhere.  Some of that code was slightly trashy, so should just be treated as a reference if it's still valuable as such.